signature/mechanism_gpgme.go

```diff
@@ -133,12 +137,56 @@ func (m *gpgmeSigningMechanism) Sign(input []byte, keyIdentity string) ([]byte,
 	if err != nil {
 		return nil, err
 	}
+
+	passphraseFile := os.Getenv(PASSPHRASE_FILE_ENVIRONMENT)
+	if options != nil && options.PassphraseFile != "" {
+		passphraseFile = options.PassphraseFile
+	}
+	if passphraseFile != "" {
+		logrus.Debugf("Reading user-specified passphrase for gpgme signing from %s", passphraseFile)
+		ppf, err := os.Open(passphraseFile)
+		if err != nil {
+			return nil, err
+		}
+		defer ppf.Close()
+
+		// Read the *first* line in the passphrase file, just as gpg(1) does.
+		buf, err := bufio.NewReader(ppf).ReadBytes('\n')
+		if err != nil && !errors.Is(err, io.EOF) {
+			return nil, fmt.Errorf("reading passphrase file: %w", err)
+		}
+		buf = append(buf, '\n') // passphrase must terminate with a new line
+
+		// Callback to write the passphrase to the specified file descriptor.
+		callback := func(uidHint string, prevWasBad bool, gpgmeFD *os.File) error {
+			if prevWasBad {
+				return fmt.Errorf("bad passphrase")
+			}
+			_, err := gpgmeFD.Write(buf)
+			return err
+		}
+		if err := m.ctx.SetCallback(callback); err != nil {
+			return nil, fmt.Errorf("setting gpgme passphrase callback: %w", err)
+		}
+
+		// Loopback mode prevents gpgme from prompting and using the callback.
```

Comment:
This reads “prevents … from … using the callback”, when we are doing it exactly to make the callback applicable.